robot_skills/src/robot_skills/robot.py

```diff
@@ -129,34 +161,136 @@ def tf_transform_pose(self, ps, frame):
-        elif isinstance(side, arms.Arm):
-            preferred_side = side
-        else:
-            print "Unknown arm side:" + str(side) + ". Defaulting to '{0}'".format(preferred_side.side)
-
-        backup_side = backup_obj_dict[preferred_side]
-        return preferred_side, backup_side
+
+    def get_robot_arm(self, required_gripper_types=None, desired_gripper_types=None,
+                            required_goals=None, desired_goals=None,
+                            required_trajectories=None, desired_trajectories=None,
+                            required_arm_name=None, desired_arm_name=None,
+                            required_objects=None, desired_objects=None):
+        """
+        Find an arm that has the needed and hopefully the desired properties.
+        Does not give an arm if all needed properties cannot be satisfied. An
+        arm returned by the call will not be able to do anything you didn't ask
+        for, except for objects held by it.
+
+        :param required_gripper_types: Collection of gripper types that must all be
+                available. None means grippers are not needed.
+        :param desired_gripper_types: Collection of gripper types where one or more
+                may be selected. None means no grippers are desired.
+
+        :param required_goals: Collection of joint goals that must all be available.
+                None means no joint goals are needed.
+        :param desired_goals: Collection of joint goals where one or more may be
+                selected. None means no joint goals are desired.
+
+        :param required_trajectories: Collection of joint trajectories that must all
+                be available. None means no joint trajectories are needed.
+        :param desired_trajectories: Collection of joint trajectories where one or
+                more may be selected. None means no joint trajectories are desired.
+
+        :param required_arm_name: Name of the arm that is needed. If set, no
+                other arm will be considered. None means any arm will do.
+        :param desired_arm_name: Name of the desired arm. This arm may or may
+                not be selected. None means there is no preference for an arm.
+
+        :param required_objects: Collection of objects that the arm must have. Special
+                pseudo-objects arms.ANY_OBJECT and arms.NO_OBJECT may be used
+                too in the collection, although they do not make much sense when used
+                together with other objects. None means there are no required objects.
+        :param desired_objects: Collection of objects that the arm may have. None
+                means there are no desired objects.
+
+        :return: An Arm of the robot with the requested properties, or None.
+                Note that the arm will never do more than you requested.
+        """
+        discarded_reasons = [] # Reasons why arms are discarded.
+
+        for arm_name, arm in self.iteritems():
```

Comment:
`self.parts.iteritems()`?

Reply:
Actually, `self.arms`, but thanks for noticing.